src/dodal/devices/linkam3.py

```diff
@@ -0,0 +1,103 @@
+
+    def __init__(self, prefix: str, name: str):
+        self.temp = epics_signal_r(float, prefix + "TEMP:")
+        self.dsc = epics_signal_r(float, prefix + "DSC:")
+        self.start_heat = epics_signal_rw(bool, prefix + "STARTHEAT:")
+
+        self.ramp_rate = epics_signal_rw(
+            float, prefix + "RAMPRATE:", prefix + "RAMPRATE:SET:"
+        )
+        self.ramp_time = epics_signal_r(float, prefix + "RAMPTIME:")
+        self.set_point = epics_signal_rw(
+            float, prefix + "SETPOINT:", prefix + "SETPOINT:SET:"
+        )
+        self.pump_control = epics_signal_r(
+            PumpControl,
+            prefix + "LNP_MODE:SET:",
+        )
+        self.speed = epics_signal_rw(
+            float, prefix + "LNP_SPEED:", prefix + "LNP_SPEED:SET:"
+        )
+
+        self.chamber_vac = epics_signal_r(float, prefix + "VAC_CHAMBER:")
+        self.sensor_vac = epics_signal_r(float, prefix + "VAC_DATA1:")
+
+        self.error = epics_signal_r(str, prefix + "CTRLLR:ERR:")
+
+        # status is a bitfield stored in a double?
+        self.status = epics_signal_r(float, prefix + "STATUS:")
+
+        self.set_readable_signals(
+            read=(self.temp,), config=(self.ramp_rate, self.speed, self.set_point)
+        )
+
+        super().__init__(name=name)
+
+    async def _move(self, new_position: float, watchers: List[Callable] = []):
+        # time.monotonic won't go backwards in case of NTP corrections
+        start = time.monotonic()
+        old_position = await self.set_point.get_value()
+        await self.set_point.set(new_position, wait=True)
+        async for current_position in observe_value(self.temp):
+            for watcher in watchers:
+                watcher(
+                    name=self.name,
+                    current=current_position,
+                    initial=old_position,
+                    target=new_position,
+                    time_elapsed=time.monotonic() - start,
+                )
+            if abs(current_position - new_position) < self.tolerance:
+                await asyncio.sleep(self.settle_time)
+                break
+
+    def set(self, new_position: float, timeout: Optional[float] = None) -> AsyncStatus:
+        watchers: List[Callable] = []
+        coro = asyncio.wait_for(self._move(new_position, watchers), timeout=timeout)
+        return AsyncStatus(coro, watchers)
+
+    # TODO: Check bitshift order
+    async def in_error(self) -> bool:
+        return bool(int(await self.status.get_value()) & 1)
+
+    async def at_setpoint(self) -> bool:
+        return bool(int(await self.status.get_value()) & 1 << 1)
+
+    async def heater_on(self) -> bool:
+        return bool(int(await self.status.get_value()) & 1 << 2)
+
+    async def pump_on(self) -> bool:
+        return bool(int(await self.status.get_value()) & 1 << 3)
+
+    async def pump_auto(self) -> bool:
+        return bool(int(await self.status.get_value()) & 1 << 4)
```

Comment:
Lots of similar code here. Extract a helper e.g. `is_nth_bit_set(n)`?

Reply:
Do you know if the bitshift order has been checked? Whilst tidying this up if I could remove the TODO if possible.

Reply:
`grep -C50 -rF "P_CtrlStatus" /dls_sw/prod/R3.14.12.7/support/linkam3/2-11/` should help to check it.

Reply:
controllerError << 0
heater1RampSetPoint << 1
heater1Started << 2
lnpCoolingPumpOn << 3
lnpCoolingPumpAuto << 4
sampleCal << 5

Looks like they're correct to the best of my reading ability: but are they required? These functions aren't being used currently. sampleCal isn't defined, but I also am not sure what it means- Sample Calibrated? Calibrated with what?

Reply:
> Are they required

Probably someone was just trying to provide a "complete" bluesky interface rather than just what was needed for the previous i22 experiment?

> sampleCal isn't defined, but I also am not sure what it means- Sample Calibrated? Calibrated with what?

[code](https://github.com/dls-controls/linkam3/blob/afa5be4056763aa71640ab7b8f5661dff3931720/linkamT96App/src/include/ControllerAPI.h#L252)

> the stage is currently using user defined sample calibration values.

i.e. a custom calibration curve between the actual sensor measurement and reported temperature value.

Reply:
https://github.com/DiamondLightSource/dodal/issues/338
Made an issue for actually implementing them in configuration/read signals, and did the refactoring.